### PR TITLE
Add support for (At/Under)MinISR-based throttling/cancellation

### LINKED ISSUE #1467: Enhance Concurrency Adjuster to take Number of AtMinISR partitions w/o offline replicas into account
Cruise Control's Concurrency Adjuster automatically sets up the inter-broker replica reassignment and leadership concurrency for executions (see https://github.com/linkedin/cruise-control/issues/1288 and https://github.com/linkedin/cruise-control/issues/1298).

The current automation is based on a feedback loop that uses selected broker metrics -- i.e. it uses the relevant broker metrics to decide whether to increase, decrease, or maintain the current pace of reassignment. However, we observed that under certain cases while broker metrics are within an acceptable range, the cluster might still have UnderMinISR partitions. This makes such partitions read-only for partitions configured with _acks=all_ -- i.e. consumers can get data from them, but such producers cannot produce to them.

Take the number of AtMinISR partitions w/o offline replicas into account while making decisions to increase, decrease, or maintain the current pace of reassignment -- e.g. if there are such partitions, recommend decreasing the concurrency multiplicatively.

### PR DESCRIPTION
**Pull request by efeg**
This PR resolves #1467, #1523, and #1524.

**1. Key Concepts:**
- **In-Sync Replicas (ISR)**:  Replicas of a partition that are "in-sync" with the leader
- **min.insync.replicas**: A topic-level config to indicate the minimum number of replicas that must acknowledge a write for the write to be considered successful when a producer sets acks to "all"
- **AtMinISR**: A partition with min.insync.replicas in its ISR
- **UnderMinISR**: A partition with fewer than min.insync.replicas in its ISR

**2. Notes**
* Introduces static configs to enable/disable (At/Under)MinISR-based throttling/cancellation. A followup patch will make (At/Under)MinISR-based throttling/cancellation dynamically configurable.
* Avoids adding an external library (e.g. `Caffeine` or `Guava`) for `TopicMinIsrCache`.
* Also contains minor refactoring in `MetadataClient` and `LoadMonitor`.

### CODE REVIEW

File: docs/wiki/User Guide/Configurations.md [hunk 1]
```diff
@@ -164,6 +164,10 @@ The following configurations are inherited from the open source Kafka client con
...
+| concurrency.adjuster.num.min.isr.check                            | Integer | N         | 5                                                                                                                                                                                                                                                                                                                                                                                                                      | The number of times that (At/Under)MinISR status of partitions in the cluster will be checked during each concurrency auto adjustment interval. For example, if the concurrency auto adjustment interval is 6 minutes and this config is 5, then (At/Under)MinISR status of partitions in the cluster will be checked once in every 72 seconds.                                |
```
Comment (zornhsu) [inline docs/wiki/User Guide/Configurations.md:167]: Nit: These settings look relevant to line 164. Maybe we can move them 2 lines up to line 165? Or perhaps we want to always append new features only to the bottom?

Comment (efeg) [reply docs/wiki/User Guide/Configurations.md:167]: As long as the config is defined within the right config group (e.g. `Executor Configurations`), I feel the row order within the table should not really matter.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/common/TopicMinIsrCache.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+      try {
+        removeOldEntriesFromCache();
+      } catch (Throwable t) {
+        LOG.warn("Received exception when trying to remove old entries from the cache.", t);
+      }
+    }
+  }
+
+  /**
+   * @return An unmodifiable view of cache , containing minIsr with time by topic name.
```
Comment (zornhsu) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/common/TopicMinIsrCache.java:71]:
```suggestion
   * @return An unmodifiable view of cache, containing minIsr with time by topic name.
```

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/executor/ConcurrencyAdjusterTest.java [hunk 3]
```diff
@@ -92,6 +102,140 @@ public static void setup() {
...
+
+    // 1.2. Leadership reassignment (non-capped)
+    recommendedConcurrency = ExecutionUtils.recommendedConcurrency(cluster,
+                                                                   minIsrWithTimeByTopic,
+                                                                   MOCK_MAX_LEADERSHIP_MOVEMENTS,
+                                                                   ConcurrencyType.LEADERSHIP);
+    assertEquals(MOCK_MAX_LEADERSHIP_MOVEMENTS / MOCK_MD_LEADERSHIP, recommendedConcurrency.intValue());
+
+    // 1.3. Inter-broker replica reassignment (capped)
+    int currentMovementConcurrency = (MOCK_MIN_PARTITION_MOVEMENTS_PER_BROKER * MOCK_MD_INTER_BROKER_REPLICA + 1) - 1;
```
Comment (zornhsu) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/executor/ConcurrencyAdjusterTest.java:174]: What is the benefit of using `(x * y +  1) - 1` instead of `x * y`?

**Comment by efeg** [comment]
@zornhsu Followup patch for this PR is available here: https://github.com/efeg/cruise-control/pull/4
I will create a new pull request under `cruise-control` repository once this PR is merged.

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/common/TopicMinIsrCacheTest.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    assertEquals(Integer.parseInt(MIN_IN_SYNC_REPLICAS_VALUE), minIsrWithTimeByTopic.get(MOCK_TOPIC).minISR());
+    assertEquals(Integer.parseInt(MIN_IN_SYNC_REPLICAS_VALUE_2), minIsrWithTimeByTopic.get(MOCK_TOPIC_2).minISR());
+    assertEquals(Integer.parseInt(MIN_IN_SYNC_REPLICAS_VALUE_3), minIsrWithTimeByTopic.get(MOCK_TOPIC_3).minISR());
+    assertEquals(MOCK_TIME_MS, minIsrWithTimeByTopic.get(MOCK_TOPIC).timeMs());
+    assertEquals(MOCK_TIME_MS, minIsrWithTimeByTopic.get(MOCK_TOPIC_2).timeMs());
+    assertEquals(MOCK_TIME_MS, minIsrWithTimeByTopic.get(MOCK_TOPIC_3).timeMs());
+
+    EasyMock.verify(_mockTime, describeConfigsResult, describedConfigsFuture, describedConfigsFuture3, topicConfig3);
+
+    // 4. Test putting one more new entry when the cache is full.
```
Comment (zornhsu) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/common/TopicMinIsrCacheTest.java:157]: Should we add one more test to make sure cache is empty after cache expire time?

**Review by zornhsu — APPROVED** [review]
